### DISCUSSION TITLE
Учет фактически существующих в команде

### DISCUSSION THREAD

**Discussion by dosinabox** [dosinabox/g2nr_unofficial_update#436] [General]
вообщем с правками что Марио и Гирион не считается мы опять сломали логику авторов. Марио и Гирион должны учитываться как в оригинальных скриптах, так как дело не только в максимальном кол-ве членов экипажа, но и в **минимальном**.

а так будет получится что герой взял капитана + 5 человек(учитывая Марио и Гириона) а уплыть не может. И капитан при этом будет говорит что ему надо пять человек.

_Originally posted by @Kor-Angar in https://github.com/dosinabox/g2nr_unofficial_update/issues/37#issuecomment-527284545_

**Comment by Kor-Angar** [comment]
Ли не дал опыта когда я позвал его на корабль
https://youtu.be/OzduAYROvw0?t=4773

**Comment by dosinabox** [comment]
Готово, нужно будет проверить.

**Comment by Kor-Angar** [comment]
там не только Ли. надо всю команду перепроверить

**Comment by dosinabox** [comment]
На самом деле всё интереснее. Опыт за любого члена команды дается только если сразу взять их в команду. Если брать время на подумать, то не дается. На примере Ли:
**Пакуй свои вещи!** - опыт есть
**Я дам тебе знать, если ты мне понадобишься.** -> **Ты мне все-таки нужен!** - опыта нет
Так же со всеми остальными.

**Comment by dosinabox** [comment]
На примере Диего:
**Почему бы тебе не отправиться в путь со мной? Встретимся в гавани.** - опыт есть
**Возможно, я дам тебе знать, когда придет время.** -> **Возвращайся. Я хочу, чтобы ты сопровождал меня.** - опыта нет

**Comment by dosinabox** [comment]
На примере Ангара:
**Иди в гавань. Встретимся там.** - опыт есть
**Возможно, я обращусь к тебе позже.** -> **Возвращайся на борт.** - опыта нет

**Comment by Kor-Angar** [comment]
гм.. может так было задумано? 

а с капитаном как? :-)

**Comment by dosinabox** [comment]
Думаю да, похоже на обдуманное решение.
А опыт за капитана дается только тогда, когда он становится капитаном, а других взять уже нельзя.

**Comment by Kor-Angar** [comment]
Может просто давать в два раза меньше опыта?

**Comment by dosinabox** [comment]
за что и почему?

**Comment by Kor-Angar** [comment]
> Думаю да, похоже на обдуманное решение.

а может потому что  **Возвращайся на борт.** - опыта нет, так как подразумевается что Герой взял его на борт, получил опыт, отказался от него и теперь повторно берет не борт.

?

**Comment by dosinabox** [comment]
Ну, можно (и, наверное, нужно) фразы менять. С опытом не знаю как быть.

**Comment by dosinabox** [comment]
> Ну, можно (и, наверное, нужно) фразы менять.

Готово.

Фразы, если персонаж **был** в команде:
- Возвращайся на борт. (Ангару)
- Возвращайся. Ты нужен мне. (Горну)
- Возвращайся. Я хочу, чтобы ты сопровождал меня. (Диего)
- Возвращайся, я не могу найти другого кузнеца. (Беннету)
- Мне казалось, я уже говорил об этом. (Биффу)
- Мудрое решение. Надеюсь, ты больше его не переменишь. (Ватрас)

Фразы, если персонаж **не был** в команде:
- Ты нужен мне. (Ангару)
- Мне нужен друг, всегда готовый помочь. (Горну)
- Мне не помешала бы твоя помощь. (Диего)
- Будь моим кузнецом. Увидимся в гавани. (Беннету)
- Мы еще поговорим. (Биффу)
- Только не трать время понапрасну. Помни, друг мой, враг не дремлет. (Ватрас)

**Comment by Kor-Angar** [comment]
> Фразы, если персонаж не был в команде:

так в этом случае опыт будет даваться ?

> С опытом не знаю как быть.

можно на форуме спросить. может подскажут

**Comment by dosinabox** [comment]
> так в этом случае опыт будет даваться ?

Пока нет. Спорно.

**Comment by Kor-Angar** [comment]
> Пока нет. Спорно.

чисто теоретически я могу за всю команду получить 0 опыта. это не спорно, это не нормально ;-)

**Comment by dosinabox** [comment]
Твое предложение - полный опыт при откладывании вопроса или половина?

**Comment by Kor-Angar** [comment]
раньше я сомневался, думал что это авторская задумка,  и в качестве "штрафа" предложил половину опыта.

НО после того как ты сделала фразы 
- Фразы, если персонаж был в команде:
- Фразы, если персонаж не был в команде:

стал склоняюсь к мысли что Пираньи просто неучли такой вариант, когда герой не сразу скажет "да" или "нет", а по сути возьмет время подумать и потом скажет "да".
таким образом теперь получается три варианта:

- сразу да
- подумал и да 
- нет

у капитана на самом деле такая же история. можно сразу сказать "да", а можно взять паузу, подумать, а потом сказать "да". но при этом опыт дают в любом случае :-)

ситуация очень редкая, за сотни прохождений, я её "ловил" буквально пару раз.


я предлагаю эту тему помусолить на форуме.

**Comment by Kor-Angar** [comment]
но отсюда вытекает новый вопрос: 
если я возьму к примеру 5 человек, возьму опыт c них, а потом их выгоню. 
потом возьми других пять, возьму опыт и их выгоню. и так далее и так далее.

таким образом я получу сверх положенного опыта? т.е. получу опыт не за фактическое членов команды.

**Comment by dosinabox** [comment]
это и в оригинале можно сделать

**Comment by dosinabox** [comment]
получается такой квест "найди всех друзей")

**Comment by Kor-Angar** [comment]
так может быть сделать счётчик и после того как герой получит опыт за такое то кол-во человек, больше опыта не давать?

**Comment by Kor-Angar** [comment]
> получается такой квест "найди всех друзей")

если взять всех НПС из Готики 1 будет еще одно лишнее место ;-)

**Comment by dosinabox** [comment]
> так может быть сделать счётчик и после того как герой получит опыт за такое то кол-во человек, больше опыта не давать?

По-моему, это лишнее. Опыт дается за дело: обойти две карты и перебрать кучу человек. Опыт в обмен на игровое время это нормально.

**Comment by dosinabox** [comment]
Фух, готово. Нужно **очень** тщательно проверить, потому что многое пришлось перелопатить. В идеале вообще всех возможных соратников перебрать, с присоединением и отсоединением. Посмотреть, кто когда обижается, кто когда атакует. Как опыт дается. Записи в дневнике. Озвучка реплик, субтитры и строки диалогов. Перемещения в гавань и на корабль.
[g2a_nr_scriptpatch_v22_hotfix_10.10.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/3714250/g2a_nr_scriptpatch_v22_hotfix_10.10.zip)
Новая игра не нужна.

**Comment by Kor-Angar** [comment]
> Фух, готово. Нужно очень тщательно проверить, потому что многое пришлось перелопатить. В идеале вообще всех возможных соратников перебрать, с присоединением и отсоединением. Посмотреть, кто когда обижается, кто когда атакует.

вроде проверял. давай всех разгоню и попробую опять собрать. напиши кого несколько раз нужно выгнать(Ли?, Волк?), что бы они окончательно обиделись на героя :-)

**********
Ссылки по теме
https://github.com/dosinabox/g2nr_unofficial_update/issues/193
https://github.com/dosinabox/g2nr_unofficial_update/issues/122
https://github.com/dosinabox/g2nr_unofficial_update/issues/38
https://github.com/dosinabox/g2nr_unofficial_update/issues/37

**Comment by Kor-Angar** [comment]
таки [несмог уплыть ](https://youtu.be/KLzt9TrjFNc?t=5983) :-D

- [x] - Мильтен (бессмертный) в команде
- [ ] - Лестер (бессмертный) ушел в отказ
- [x] - Диего (бессмертный) в команде
- [ ] - Ли (бессмертный) ушел в отказ
- [x] - Беннет (бессмертный) в команде
- [ ] - Ларес (бессмертный) ушел в отказ
- [ ] - остальные НПС погибли

Сейв
https://yadi.sk/d/8UGOonInboMuGg

**Reply by Kor-Angar** [reply]
> Возможна ли такая ситуация в оригинале Г2НВ?

да

**Reply by Kor-Angar** [reply]
после просмотра карты можно набирать команду в 5 главе. код карты ItWr_Seamap_Irdorath

**Всегда плывут + бессмертные**
-  Мильтен (бессмертный) в команде
-  Диего (бессмертный) в команде
-  Беннет (бессмертный) в команде

**Могут, отказаться плыть**
-  Лестер (бессмертный)
-  Ли (бессмертный)
-  Ларес (бессмертный)
-  Волк
-  Марио
-  Ватрас

Всегда плывут, но смертные 
- Ангару // высокая вероятность что погиб в 4 главе
- Горну // высокая вероятность что погиб в 4 главе

Прочие смертные(проверить)
- Бифф // высокая вероятность что погиб в 4 главе
- Гирион

**Reply by dosinabox** [reply]
что с этим можно сделать?

**Reply by Kor-Angar** [reply]
можно сделать проверку на смерть потенциальных "безотказных" членов команды. если много мертвых то Лестеру, Ли и Ларесу запрещать отказываться.

то есть это:
if Crewmember_Count < Min_Crew
&& смерть Ангар
&& смерть Горн
&& смерть Бифф

**Reply by Kor-Angar** [reply]
а вот кстати альтернативный вариант :-)

> может добавить возможность вернуть Лестера, Ли, Лареса,Волка и Марио в команду через заклинание "забыть"?

**Reply by dosinabox** [reply]
> может добавить возможность вернуть Лестера, Ли, Лареса,Волка и Марио в команду через заклинание "забыть"?

```cpp
if(Lares_IsOnBoard == LOG_FAILED)
{
	Lares_IsOnBoard = FALSE;
};
```

```cpp
if(Wolf_IsOnBoard == LOG_FAILED)
{
	Wolf_IsOnBoard = FALSE;
};
```

```cpp
if(Lee_IsOnBoard == LOG_FAILED)
{
	Lee_IsOnBoard = FALSE;
};
Lee_Nerver = 0;
```

```cpp
if(Lester_IsOnBoard == LOG_FAILED)
{
	Lester_IsOnBoard = FALSE;
};
Lester_Nerver = 0;
```

```cpp
if(Mario_IsOnBoard == LOG_FAILED)
{
	Mario_IsOnBoard = FALSE;
};
Mario_Nerver = 0;
```

**Comment by dosinabox** [comment]
Должен ли счетчик членов команды уменьшаться при смерти кого-либо смертного из команды?

**Reply by Kor-Angar** [reply]
уточни, про что именно ты говоришь. на сколько я понял после того как НПС становится членом команды, он становится бессмертным(по крайней мере так было у Горна и Ангара).

Т.е. в итоге счётчик нам нужен для того что бы в сухом остатке на корабле на Ирдорате был капитан + пять человек + герой.

**Reply by Kor-Angar** [reply]
> после того как НПС становится членом команды, он становится бессмертным

это тоже надо перепроверить на всех НПС.

**Reply by Kor-Angar** [reply]
@dosinabox  ты же бессмертие делаешь в рамках общей функции ? там по моему и хп увеличивается на 1 и флаг бессмертия ставится?

**Reply by dosinabox** [reply]
> ты же бессмертие делаешь в рамках общей функции ? там по моему и хп увеличивается на 1 и флаг бессмертия ставится?

Да. Вопрос был задан до создания этой общей функции и сейчас уже не актуален.

**Reply by dosinabox** [reply]
https://github.com/dosinabox/g2nr_unofficial_update/blob/work/PrjGOTHIC/Story/B_Content/B_JoinShip.d

**Reply by Kor-Angar** [reply]
спс

**Reply by Kor-Angar** [reply]
вместо горы проверок  по **_WasOnBoard , лучше сделать одну по какому нибудь аивер НПС. тогда и кода будет меньше и переменных тоже.

```D
		if(slf.aivar[****] != 666)
		{
			B_GivePlayerXP(XP_Crewmember_Success);
                        slf.aivar[****] = 666
		};
```

или по атрибуту :-)

```D
	if(slf.lp != 666)
		{
			B_GivePlayerXP(XP_Crewmember_Success);
                        slf.lp = 666
		};
```

**Comment by Kor-Angar** [comment]
смотрю не забыл ли я про кого нибудь случайно :-)
**MIS_ReadyforChapter6**
![image](https://user-images.githubusercontent.com/49514764/74334763-9606dd00-4dab-11ea-9828-707a0f6c5486.png)

**Crewmember_Count**
![image](https://user-images.githubusercontent.com/49514764/74334805-ac149d80-4dab-11ea-86a6-79cf9fc46628.png)

**Comment by Kor-Angar** [comment]
> Возможна ли такая ситуация в оригинале Г2НВ?
> 
> проверить?
> 
> было бы шикарно

тогда предлагаю такой сценарий

**Убиваю в 4 главе:**
- Ангару
- Горну
- Бифф

**Беру с собой**
- Мильтен (бессмертный) всегда  в команде
- Диего (бессмертный) всегда  в команде
- Беннет (бессмертный) всегда в команде
- Лестер (бессмертный)
- Ли (бессмертный)
- Ларес (бессмертный)
- Волк
- Марио
- Ватрас
- Гирион

Делаю контрольный сейв и смотрю чем дело заканчивается -)

**Comment by Kor-Angar** [comment]
И Ватрас надо будет еще отдельно проверить, когда он находясь в команде посылает героя и когда он уже поднявшись на борт посылает героя 😀

**Reply by Kor-Angar** [reply]
@dosinabox  что то про это мы в дискорде на днях перпеписывались. продублируй тут свежую правку про ватраса.

**Reply by dosinabox** [reply]
какую именно?

**Comment by Kor-Angar** [comment]
не нашел топик про Гириона в команде. пишу здесь свои правки https://github.com/Kor-Angar/G2_NotR_Kor_Angar/commit/1c90964afae279bd9cb1c79bb6474266dce1df74 Правка простая: добиться что бы Гирион появлялся в порту вовремя  и не опаздывал на корабль :-)

если в порту есть пару человек то Гириону меняется расписание в  B_JoinShip.d
```D
	if 	Crewmember_Count > 2
	{
		B_StartOtherRoutine(Girion,"WaitForShip");
	};
```

у Гириона проверку на MIS_ShipIsFree   из Condition перенёс в реплики.
```D
func int DIA_Girion_CATCHPLAYERSTOLENSHIP_Condition()
{
	if	(Kapitel >= 5) 
	//&&	(MIS_ShipIsFree == TRUE) 
	&&	(Npc_GetDistToWP(self,"NW_CITY_WAY_TO_SHIP_25") < 1000)
	{
		return TRUE;
	};
};

func void DIA_Girion_CATCHPLAYERSTOLENSHIP_Info()
{
	AI_Output(self,other,"DIA_Girion_CATCHPLAYERSTOLENSHIP_08_00");	//Эй! Ты можешь сказать мне, что ты делаешь здесь?
	Info_ClearChoices(DIA_Girion_CATCHPLAYERSTOLENSHIP);
	if	(MIS_ShipIsFree == TRUE) 
	{
		Info_AddChoice(DIA_Girion_CATCHPLAYERSTOLENSHIP,"Я не знаю, о чем ты говоришь.",DIA_Girion_CATCHPLAYERSTOLENSHIP_no);
	};
	Info_AddChoice(DIA_Girion_CATCHPLAYERSTOLENSHIP,"Прочь с моей дороги, или мне придется убить тебя.",DIA_Girion_CATCHPLAYERSTOLENSHIP_weg);
	Info_AddChoice(DIA_Girion_CATCHPLAYERSTOLENSHIP,"Мне нужен этот корабль. Поэтому я забираю его.",DIA_Girion_CATCHPLAYERSTOLENSHIP_ship);
};

func void DIA_Girion_CATCHPLAYERSTOLENSHIP_no()
{
	AI_Output(other,self,"DIA_Girion_CATCHPLAYERSTOLENSHIP_no_15_00");	//Я не знаю, о чем ты говоришь.
	AI_Output(self,other,"DIA_Girion_CATCHPLAYERSTOLENSHIP_no_08_01");	//Я говорю о том, что ты сделал с корабельной стражей. Это дело плохо пахнет.
	if((hero.guild == GIL_KDF) && !Npc_IsDead(Schiffswache_212) && !Npc_IsDead(Schiffswache_213))
	{
		AI_Output(self,other,"DIA_Girion_CATCHPLAYERSTOLENSHIP_no_08_02");	//Даже несмотря на то, что ты маг, я не доверяю тебе.
	};
};


var int Girion_WantstoKillSC;

func void DIA_Girion_CATCHPLAYERSTOLENSHIP_weg()
{
	AI_Output(other,self,"DIA_Girion_CATCHPLAYERSTOLENSHIP_weg_15_00");	//Прочь с моей дороги, или мне придется убить тебя.
	if	!Npc_IsDead(Schiffswache_212) 
	&&	!Npc_IsDead(Schiffswache_213)
	&&	MIS_ShipIsFree == TRUE
	{
		AI_Output(self,other,"DIA_Girion_CATCHPLAYERSTOLENSHIP_weg_08_01");	//Возможно, тебе удалось обмануть охрану корабля, но со мной тебе так просто не справиться, мой маленький друг.
		if(hero.guild == GIL_PAL)
		{
			AI_Output(self,other,"DIA_Girion_CATCHPLAYERSTOLENSHIP_weg_08_02");	//Даже хотя ты один из нас, это не дает тебе права воровать собственность короля. Умри, проклятый предатель.
		};
	}
	else
	{
		AI_Output(self,other,"DIA_Girion_Verrat_08_00");	//Предатель!
	};
	AI_StopProcessInfos(self);
	B_Attack(self,other,AR_KILL,1);
	Girion_WantstoKillSC = TRUE;
};
```
Гирион приходит в порт и в ранних главах, если герой проник на корабль. что он там будет делать я пока не придумал. при следующем прохождение  буду думать :-)
```D
MIS_ShipIsFree = TRUE;
	B_CheckLog();
	if(Kapitel == 5)
	{
		B_StartOtherRoutine(Girion,"WaitForShip");
	}
	else
	{
		B_StartOtherRoutine(Girion,"WaitFor");
	};
```

**Comment by Kor-Angar** [comment]
```D
	if (MIS_OCGateOpen == TRUE)
	&& (Pal_Schiffswache_Exchange_onetime == FALSE)
	{
			B_StartOtherRoutine	(Pal_212_Schiffswache,"ShipFree");
			B_StartOtherRoutine	(Pal_213_Schiffswache,"ShipFree");
			Pal_Schiffswache_Exchange_onetime = TRUE;
			MIS_ShipIsFree = TRUE;
	};
```

а где у тебя находится такая конструкция? я так понимаю что это был альтернативный способ получить доступ на корабль. только мен кажется что такая конструкция должна находиться у Хагена в диалоге про открытие ворот в замке.

**Reply by dosinabox** [reply]
так и есть:
```cpp
func void DIA_Lord_Hagen_GateOpen_Info()
{
	AI_Output(other,self,"DIA_Lord_Hagen_Add_15_29");	//Орки ворвались в замок в Долине Рудников!
	AI_Output(self,other,"DIA_Lord_Hagen_Add_04_30");	//О, Иннос! Что именно там произошло?
	AI_Output(other,self,"DIA_Lord_Hagen_Add_15_31");	//Почему-то ворота оказались открытыми...
	AI_Output(self,other,"DIA_Lord_Hagen_Add_04_32");	//Почему-то?! Но как это возможно... В замке наверняка есть предатель!
	B_StartOtherRoutine(Schiffswache_212,"ShipFree");
	B_StartOtherRoutine(Schiffswache_213,"ShipFree");
	B_StartOtherRoutine(Girion,"WaitForShip");
	MIS_ShipIsFree = TRUE;
	B_CheckLog();
};
```

**Reply by Kor-Angar** [reply]
тут же Pal_212_Schiffswache и Schiffswache_213 надо менять fight_tactic

**Comment by Kor-Angar** [comment]
@dosinabox 

> квест "Моя команда" больше подходит в раздел общей информации, как думаешь?

**Reply by Kor-Angar** [reply]
Для ирдората подходит, а для 5 главы не уверен что подойдёт. можно сделать второй топик в общей информации и там дублировать информацию. тогда тот что миссия  уйдёт в архив, а новый будет в общей инфе.
второй топик лучше создать на ирдорате указав по факту кто есть на корабле и чем он поможет герою, это исключить "мёртвые души".

**Reply by dosinabox** [reply]
5 глава - "Невозможно плыть без команды" (Topic_Crew)
6 глава - "Моя команда" (TOPIC_MyCrew)

**Reply by Kor-Angar** [reply]
гуд

**Reply by Kor-Angar** [reply]
так оно сейчас и так есть

**Reply by dosinabox** [reply]
> второй топик лучше создать на ирдорате указав по факту кто есть на корабле и чем он поможет герою, это исключить "мёртвые души".

Это и есть "Моя команда" (TOPIC_MyCrew). Единственное что я предлагаю - создавать его в разделе общей информации, а не в текущих заданиях.

**Reply by Kor-Angar** [reply]
ну это было бы логично :-)

**Comment by Kor-Angar** [comment]
после того как все взашли на борт , можно снять флаги бессмертия с Лареса, Ли , Ватраса, Беннета если они не на борту
